### PR TITLE
@types/jsonwebtokens - Allow Generics for synchronous verify

### PR DESCRIPTION
**Pull request by zRelux**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [this](https://github.com/auth0/node-jsonwebtoken/issues/483)

### CODE REVIEW

File: types/jsonwebtoken/index.d.ts [hunk 1]
```diff
@@ -169,7 +169,7 @@ export function sign(
  * [options] - Options for the verification
  * returns - The decoded token.
  */
-export function verify(token: string, secretOrPublicKey: Secret, options?: VerifyOptions): object | string;
+export function verify<T>(token: string, secretOrPublicKey: Secret, options?: VerifyOptions): T | string;
```
Comment (peterblazejewicz) [inline types/jsonwebtoken/index.d.ts]: This probably should read:
```suggestion
export function verify<T = object>(token: string, secretOrPublicKey: Secret, options?: VerifyOptions): T | string;
```
to stay in sync with previous implementation (backward comp.):
```ts
jwt.verify(token, cert, { audience: [/urn:f[o]{2}/, "urn:bar"] }, (err, decoded) => {
    // if audience mismatch, err == invalid audience
    const result = decoded as TestObject;
});
```
while providing all benefits of generics in a rewritten test case:
```ts
jwt.verify<TestObject>(token, getKey, (err, decoded) => {
    const result = decoded;
    console.log(result.foo); // bar
});
```

**Review by peterblazejewicz — COMMENTED** [review]
Please amend the test to apply this change to some scenarios.
What about async version of `verify`? What about generic verify callback like:
```ts
export type VerifyCallback<T> = (
    err: VerifyErrors,
    decoded: T,
) => void;
¬``

File: types/jsonwebtoken/index.d.ts (hunk 1 above)
Comment (LinusU) [reply types/jsonwebtoken/index.d.ts]: Hmmm, shouldn't this be `T extends object`? I don't think that the library can return anything else than an object in that place (e.g. `number`, `boolean`, etc.)

**Comment by LinusU** [comment]
Hmm, I don't think that this should be merged since it goes against the last rule here:

https://github.com/Microsoft/dtslint/blob/master/docs/no-unnecessary-generics.md

Also, I think that using this would kind of be an anti pattern since there is no guarantee at all that `T` is actually returned from the `decode` function 🤔

**Comment by zRelux** [comment]
Sorry that I left this behind but have been quite busy with work.
So the issue I faced and it seems that other people are facing is when we do something like this
```typescript
const decodedToken = jwt.verify(token, secret);

// TYPE error email not present
decodedToken.email
```
My reasoning for this PR was that being that we are verifying we know what kind of payload is present in the token hence a generics would be ideal to resolve the type issue.

I do agree that the extends object is needed as it would make it backwards compatible.

**Comment by LinusU** [comment]
> My reasoning for this PR was that being that we are verifying we know what kind of payload is present in the token hence a generics would be ideal to resolve the type issue.

In that code, I don't see that the payload is verified, if you verify with a function that either returns `_ is T` or `asserts _ is T` then TypeScript should change the type of the payload after that function call, e.g.

```js
interface Token { email: string }

function verifyPayload (payload: any): asserts payload is Token {
  if (typeof payload !== 'object' || typeof payload.email !== 'string') throw new Error('...')
}

// ...

const decodedToken = jwt.verify(token, secret)

// TYPE error email not present
decodedToken.email

verifyPayload(decodedToken)

// Here everything works:
decodedToken.email
```

-----

If you know that you will only be passed valid tokens, you can simply cast the returned value from the function:

```js
interface Token { email: string }

const decodedToken = jwt.verify(token, secret) as Token
```

This is the suggestion that was given in the guidelines I linked, and is the preferred way of doing this instead of adding a generic parameter.

In fact, I think we should consider changing the return type of `verify` to `unknown` 🤔

**Pull request closed without merging**